### DISCUSSION TITLE
Counter-Strike 2 support

### DISCUSSION THREAD

**Discussion by justMemou** [saul/demofile#384] [Questions] [answered]
Hey Saul,

I know it's a bit early to say, but are you planning to upgrade the tool for Counter-Strike 2?

Thanks

**Comment by BlackYuzia** [comment]
I try record demo on `47/13858 964` cs2 game version.
Anyway, I can't `playdemo`. Suppose this version isn't complete for demo record/play?

[cs2demofile.zip](https://github.com/saul/demofile/files/11068790/cs2demofile.zip)

**Comment by saul** [comment]
Hi

Unfortunately updating this library to work with Source 2 would involve a practical rewrite. The API (like the events) that you listen to would all change, the properties available on entities have completely changed, etc. So any code that you have written to work with the current demofile library would have to be rewritten too.

With that in mind, I have decided to start a new parser called DemoFile.Net. It will be written in C#/.NET Core (the language that I use in my day job), and I'm confident I can make it significantly faster than the current Node.js parser. I'm currently working on it in my free time, but I hope to have something out by the time CS2 is released publicly. Thankfully there are already numerous Source 2 demo parsers in other languages (like Go and Java) so a lot of the hard work has already been done.

I look forward to showing you something soon!

Thanks
Saul

**Reply by BlackYuzia** [reply]
Will be compatibility with nodejs?

**Reply by trytuna** [reply]
Did you get any info about the new file format yet? As far as I can tell there are not resources online. Neither at the ValveSoftware GitHub account nor in the developer wiki from Valve.

**Reply by saul** [reply]
There will not be compatibility with Node.js, as the new library will be C# (so can only be consumed from .NET languages). I am planning to make a tool that exports information from the demo as [JSON lines](https://jsonlines.org/examples/). This will make it easy to consume the data from any language.

As for information on the 'new' file format (it's actually over 10 years old), see:

- https://developer.valvesoftware.com/wiki/Dota_2_Demo_Format (demoinfo2 docs)
- https://github.com/mitsuhiko/dota2-demoinfo2 (official demoinfo2 source)
- https://github.com/dotabuff/manta (a complete Dota 2 parser in Go)

**Reply by n1c** [reply]
Interesting idea about the JSON lines! Makes me think about what Markus did with https://github.com/markus-wa/cs-demo-minifier (on top of his own golang parser https://github.com/markus-wa/demoinfocs-golang).

**Reply by saul** [reply]
Hi all - an initial version of the C# library is now available. Please see https://github.com/saul/demofile/discussions/384#discussioncomment-7216722. Many thanks

**Comment by in0finite** [comment]
Hi @saul ,
Thank you for your answers.

Are there any news on new .NET library ? It's been almost 4 months since beta release of CS2, and there is still no tool to properly parse Demo files.
The most complete one that I found is https://github.com/markus-wa/demoinfocs-golang, however it's written in Golang, and doesn't fit perfectly into my environment.

There is also some work done here: https://github.com/skadistats/clarity/tree/csgo2 in it's `csgo2` branch.

I would also be eager to contribute to new .NET parser.

**Reply by saul** [reply]
There is no more news yet I'm afraid. I'm loathe to spend too much time working on it in while CS2 is still in private beta for a couple of reasons:

- There isn't huge demand _yet_ (e.g. no esports competitions at the moment), obviously this will come post-public release
- I don't have beta access myself, so recording demos for testing is quite difficult
- The format is very likely to change, meaning some work could be wasted

With all that said the code I have at the moment can parse demos fully (including game events/other netmessages and entity properties). The main thing that is missing is the infrastructure for entities themselves, e.g. the classes in https://github.com/saul/demofile/tree/master/src/entities.

There are no insurmountable problems here, it just requires a week or two of on/off dev time. Also I'm hoping to make the library extremely fast - .NET is my bread and butter! Don't worry - the library will come :)

**Reply by dbalders** [reply]
Hi saul, I appreciate all the work on demofile through the years, thank you for it. I wanted to ask if you had any update on your cs2 version. If not all good, just wanted to check as I'm trying to figure out how to transition. Thank you again.

**Reply by saul** [reply]
@dbalders yes, an initial version of the C# library is now available. Please see https://github.com/saul/demofile/discussions/384#discussioncomment-7216722. Many thanks

**Comment by saul** [comment, marked as answer]
Note all that are interested in CS2 support: I have just released the code for a CS2 demo parser here: https://github.com/saul/demofile-net.

It is written in C# and it's blazing fast. It reads through a full competitive game in 1.5 secs on my M1 MacBook Pro. This is significantly faster than the Node.js parser could hope to achieve.

Note that DemoFile.Net hasn't been published to NuGet yet, but I will do this weekend. I'm still trying to work out the remaining kinks, but for those interested in trying the bleeding edge can checkout the repo and give it a shot.

I'll soon be archiving this repo, so any discussion should move over to the [new repo](https://github.com/saul/demofile-net).

Thank you all for your patience.

**Comment by maxpain** [comment]
@saul what if I don't know C# and only know JavaScript?

**Reply by saul** [reply]
I'm afraid the new version of the library is .NET only, so you'd either have to learn C# or find a different library.
